File: frontend/src/radio-editions.ts
import type { Item } from './types';
import type { DayPlaylistGroup } from './day-playlists';

export function radioEdition(item: Item) {
  let tags: string[] = [];
  try { const parsed = JSON.parse(item.tags || '[]'); if (Array.isArray(parsed)) tags = parsed.filter((v): v is string => typeof v === 'string'); } catch { /* Legacy tags are optional. */ }
  const local = new Date(((item.publish_time || item.created_at || 0) + 8 * 3600) * 1000);
  const date = tags.find(t => /^radio:date:\d{4}-\d{2}-\d{2}$/.test(t))?.slice(11) || local.toISOString().slice(0, 10);
  const edition = tags.includes('radio:edition:morning') ? 'morning' : tags.includes('radio:edition:evening') ? 'evening' : local.getUTCHours() < 12 ? 'morning' : 'evening';
  return { key: `${date}:${edition}`, date, edition };
}

export function buildRadioEditions(items: Item[]): DayPlaylistGroup<Item>[] {
  const buckets = new Map<string, Item[]>();
  for (const item of items) {
    const { key } = radioEdition(item);
    const bucket = buckets.get(key);
    if (bucket) bucket.push(item); else buckets.set(key, [item]);
  }
  return [...buckets.entries()].map(([key, members]) => {
    const { date, edition } = radioEdition(members[0]);
    const startMs = Date.parse(`${date}T00:00:00+08:00`);
    const label = new Intl.DateTimeFormat('zh-CN', { timeZone: 'Asia/Shanghai', year: 'numeric', month: 'long', day: 'numeric', weekday: 'long' }).format(new Date(startMs));
    const programs = members.filter(i => isRadioProgram(i) && i.audio_url?.trim());
    const ordered = [...(programs.length ? programs : members)].sort((a,b) => (a.publish_time || a.created_at || 0) - (b.publish_time || b.created_at || 0) || a.id.localeCompare(b.id));
    const playable = ordered.filter(i => i.audio_url?.trim());
    return {key, startMs, title: `${edition === 'morning' ? '早间新闻' : '晚间新闻'} · ${label}`, shortTitle: edition === 'morning' ? '早间新闻' : '晚间新闻', items: ordered, itemIds: ordered.map(i=>i.id), playbackIds: playable.map(i=>i.id), playableCount: playable.length, totalDurationSec: playable.reduce((n,i)=>n+Math.max(0,i.duration_sec || 0),0)};
  }).sort((a,b)=>b.startMs-a.startMs || a.key.localeCompare(b.key));
}

export function isRadioProgram(item: Item): boolean {
  try { const tags = JSON.parse(item.tags || '[]'); return Array.isArray(tags) && tags.includes('radio:program'); } catch { return false; }
}

export function programSectionCount(item: Item): number {
  try { const tags = JSON.parse(item.tags || '[]'); return Array.isArray(tags) ? Number(tags.find((t: unknown) => typeof t === 'string' && /^radio:sections:\d+$/.test(t))?.split(':')[2] || 0) : 0; } catch { return 0; }
}

File: android_client/lib/radio_editions.dart
import 'dart:convert';
import 'day_playlist.dart';
import 'src/rust/models.dart';

({String key, String date, String edition}) radioEdition(Item item) {
  List<String> tags = [];
  try {
    final parsed = jsonDecode(item.tags ?? '[]');
    if (parsed is List) tags = parsed.whereType<String>().toList();
  } on FormatException {
    /* Historical tags can be plain text. */
  }
  final local = DateTime.fromMillisecondsSinceEpoch(
    (item.publishTime ?? item.createdAt ?? 0) * 1000,
    isUtc: true,
  ).add(const Duration(hours: 8));
  final dates = tags.where(
    (t) => RegExp(r'^radio:date:\d{4}-\d{2}-\d{2}$').hasMatch(t),
  );
  final date = dates.isNotEmpty
      ? dates.first.substring(11)
      : local.toIso8601String().substring(0, 10);
  final edition = tags.contains('radio:edition:morning')
      ? 'morning'
      : tags.contains('radio:edition:evening')
      ? 'evening'
      : local.hour < 12
      ? 'morning'
      : 'evening';
  return (key: '$date:$edition', date: date, edition: edition);
}

List<DayPlaylistGroup<Item>> buildRadioEditions(Iterable<Item> items) {
  final buckets = <String, List<Item>>{};
  for (final item in items) {
    (buckets[radioEdition(item).key] ??= []).add(item);
  }
  final result =
      buckets.entries.map((entry) {
        final meta = radioEdition(entry.value.first);
        final date = DateTime.parse('${meta.date}T00:00:00Z');
        const weekdays = ['星期一', '星期二', '星期三', '星期四', '星期五', '星期六', '星期日'];
        final label = meta.edition == 'morning' ? '早间新闻' : '晚间新闻';
        final programs = entry.value.where((i) => isRadioProgram(i) && (i.audioUrl?.trim().isNotEmpty ?? false)).toList();
        final ordered = [...(programs.isNotEmpty ? programs : entry.value)]
          ..sort((a, b) {
            final delta = (a.publishTime ?? a.createdAt ?? 0).compareTo(
              b.publishTime ?? b.createdAt ?? 0,
            );
            return delta == 0 ? a.id.compareTo(b.id) : delta;
          });
        final playable = ordered
            .where((i) => i.audioUrl?.trim().isNotEmpty ?? false)
            .toList();
        return DayPlaylistGroup<Item>(
          key: entry.key,
          dayStart: date,
          title:
              '$label · ${date.year}年${date.month}月${date.day}日 ${weekdays[date.weekday - 1]}',
          shortTitle: label,
          items: ordered,
          itemIds: ordered.map((i) => i.id).toList(),
          playbackIds: playable.map((i) => i.id).toList(),
          playableCount: playable.length,
          totalDurationSec: playable.fold(
            0,
            (n, i) => n + ((i.durationSec ?? 0) > 0 ? i.durationSec! : 0),
          ),
        );
      }).toList()..sort((a, b) {
        final delta = b.dayStart.compareTo(a.dayStart);
        return delta == 0 ? a.key.compareTo(b.key) : delta;
      });
  return result;
}

bool isRadioProgram(Item item) {
  try { final tags=jsonDecode(item.tags ?? '[]'); return tags is List && tags.contains('radio:program'); } on FormatException { return false; }
}
int programSectionCount(Item item) {
  try { final tags=jsonDecode(item.tags ?? '[]'); if(tags is List) { for(final tag in tags.whereType<String>()) { if(tag.startsWith('radio:sections:')) return int.tryParse(tag.substring(15)) ?? 0; } } } on FormatException { return 0; }
  return 0;
}
